File: app/src/main/java/net/squanchy/service/repository/injection/RepositoryModule.java
package net.squanchy.service.repository.injection;

import net.squanchy.service.DaysRepository;
import net.squanchy.service.firebase.FirebaseDbService;
import net.squanchy.service.firestore.FirestoreDbService;
import net.squanchy.service.repository.EventRepository;
import net.squanchy.service.repository.SpeakerRepository;
import net.squanchy.service.repository.firebase.FirebaseDaysRepository;
import net.squanchy.service.repository.firebase.FirebaseEventRepository;
import net.squanchy.service.repository.firebase.FirestoreSpeakerRepository;
import net.squanchy.support.lang.Checksum;

import dagger.Module;
import dagger.Provides;

@Module
public class RepositoryModule {

    @Provides
    EventRepository eventService(FirebaseDbService dbService, FirestoreDbService firestoreDbService, Checksum checksum) {
        return new FirebaseEventRepository(dbService, firestoreDbService, checksum);
    }

    @Provides
    SpeakerRepository speakerRepository(FirestoreDbService dbService, Checksum checksum) {
        return new FirestoreSpeakerRepository(dbService, checksum);
    }

    @Provides
    DaysRepository daysRepository(FirebaseDbService dbService) {
        return new FirebaseDaysRepository(dbService);
    }
}

File: app/src/main/java/net/squanchy/service/repository/injection/RepositoryModule.kt
package net.squanchy.service.repository.injection

import dagger.Module
import dagger.Provides
import net.squanchy.schedule.TracksRepository
import net.squanchy.schedule.firestore.FirestoreTracksRepository
import net.squanchy.service.firebase.FirestoreDbService
import net.squanchy.service.repository.EventRepository
import net.squanchy.service.repository.SpeakerRepository
import net.squanchy.service.repository.firestore.FirestoreEventRepository
import net.squanchy.service.repository.firestore.FirestoreSpeakerRepository
import net.squanchy.support.checksum.Checksum
import net.squanchy.support.checksum.ChecksumModule

@Module(includes = [ChecksumModule::class])
class RepositoryModule {

    @Provides
    internal fun eventRepository(firestoreDbService: FirestoreDbService, checksum: Checksum): EventRepository =
        FirestoreEventRepository(firestoreDbService, checksum)

    @Provides
    internal fun speakerRepository(dbService: FirestoreDbService, checksum: Checksum): SpeakerRepository =
        FirestoreSpeakerRepository(dbService, checksum)

    @Provides
    internal fun tracksRepository(dbService: FirestoreDbService, checksum: Checksum): TracksRepository =
        FirestoreTracksRepository(dbService, checksum)
}
